### PR TITLE
partial merge #17398: Update leveldb to 1.22+

### PR DESCRIPTION
**Pull request by kittywhiskers**
_Depends on the merger of #4242_

### CODE REVIEW

File: contrib/devtools/copyright_header.py [hunk 1]
```diff
@@ -41,7 +41,7 @@
 ]
 EXCLUDE_COMPILED = re.compile('|'.join([fnmatch.translate(m) for m in EXCLUDE]))
 
-INCLUDE = ['*.h', '*.cpp', '*.cc', '*.c', '*.py']
+INCLUDE = ['*.h', '*.cpp', '*.cc', '*.c', '*.mm', '*.py', '*.sh', '*.bash-completion']
```
Comment (UdjinM6) [inline contrib/devtools/copyright_header.py:44]: d596e3fd1f10a007f17a31e46256046cc7add567: this belongs to another PR (17052) + should also backport 14785 first I think (and add missing changes from 17398 on top of it)

File: configure.ac [hunk 2]
```diff
@@ -955,6 +975,50 @@ AC_COMPILE_IFELSE([AC_LANG_PROGRAM([[#include <fcntl.h>]],
 )
 
 # Check for reduced exports
```
Comment (UdjinM6) [inline configure.ac:977]:
```suggestion
dnl crc32c platform checks
```

**Review by UdjinM6 — CHANGES_REQUESTED** [review]

**Comment by PastaPastaPasta** [comment]
Please rebase and fix linter failure

**Comment by UdjinM6** [comment]
14128 was just merged via #4227 so linter should be happy after rebase

**Comment by kittywhiskers** [comment]
Rebased and updated

File: contrib/devtools/copyright_header.py (hunk 1 above)
Comment (UdjinM6) [reply contrib/devtools/copyright_header.py:40]: hmm... `*.mm` was added in 16314 actually

**Review by UdjinM6 — CHANGES_REQUESTED** [review]
Looks like we also need (at least) 16314. Also, I just realilzed that I don't understand how to merge this - we are mixing per-PR vs per-commit backports here. Let's move all unrelated backports (14785, 16314, 17052) into their own PR and keep this one for commits from 17398 only.

**Comment by kittywhiskers** [comment]
14785, 16314, 17052 are in their own PR (they're repeated here as the PR won't be complete otherwise) and upon merge the diffs will only be related to LevelDB

File: configure.ac [hunk 3]
```diff
@@ -559,8 +559,7 @@ case $host in
        AC_MSG_ERROR("windres not found")
      fi
 
-     CPPFLAGS="$CPPFLAGS -D_MT -DWIN32 -D_WINDOWS -DBOOST_THREAD_USE_LIB"
-     LEVELDB_TARGET_FLAGS="-DOS_WINDOWS"
+     CPPFLAGS="$CPPFLAGS -D_MT -DWIN32 -D_WINDOWS -DBOOST_THREAD_USE_LIB -D_WIN32_WINNT=0x0601"
```
Comment (UdjinM6) [inline configure.ac:582]: 29283b67a82975ef195e43af42bae6cbe293356d: CPPFLAGS change is from 14922 which looks unrelated
```suggestion
     CPPFLAGS="$CPPFLAGS -D_MT -DWIN32 -D_WINDOWS -DBOOST_THREAD_USE_LIB"
```

File: configure.ac [hunk 4]
```diff
@@ -638,35 +636,10 @@ case $host in
      ;;
    *android*)
      dnl make sure android stays above linux for hosts like *linux-android*
-     LEVELDB_TARGET_FLAGS="-DOS_ANDROID"
+     TARGET_OS=android
```
Comment (UdjinM6) [inline configure.ac:659]: 29283b67a82975ef195e43af42bae6cbe293356d: `TARGET_OS=android` is from 16110, which is complex and is not needed for leveldb backport I think
```suggestion
```

File: Makefile.am [hunk 5]
```diff
@@ -173,7 +173,17 @@ $(BITCOIN_TX_BIN): FORCE
...
+	-p "/usr/include/" \
+	-p "/usr/lib/" \
+	-p "/usr/lib64/" \
+	-p "src/leveldb/" \
+	-p "src/crc32c/" \
+	-p "src/bench/" \
+	-p "src/univalue" \
+	-p "src/crypto/ctaes" \
+	-p "src/secp256k1" \
+	-p "depends"
```
Comment (UdjinM6) [inline Makefile.am:186]: ab059c85af1a34169bb0700edcf472e452048929: some parts are from 17647
```suggestion
LCOV_FILTER_PATTERN=-p "/usr/include/" -p "/usr/lib/" -p "src/leveldb/" -p "src/crc32c/" -p "src/bench/" -p "src/univalue" -p "src/crypto/ctaes" -p "src/secp256k1"
```

**Review by UdjinM6 — CHANGES_REQUESTED** [review]

File: configure.ac (hunk 3 above)
Comment (kittywhiskers) [reply configure.ac:582]: Chalked it up to change in LevelDB's build system so ported it over

Comment (UdjinM6) [reply configure.ac:582]: https://github.com/UdjinM6/dash/commit/453b11274da379da5a653f1378cfac059f8b747e -> https://gitlab.com/UdjinM6/dash/-/pipelines/336015433 I do not see any issues, CI is happy

**Review by UdjinM6 — APPROVED** [review]
LGTM, utACK

File: test/lint/README.md [hunk 6]
```diff
@@ -19,6 +19,7 @@ maintained:
 * for `src/leveldb`: https://github.com/bitcoin-core/leveldb.git (branch bitcoin-fork)
 * for `src/univalue`: https://github.com/bitcoin-core/univalue.git (branch master)
 * for `src/crypto/ctaes`: https://github.com/bitcoin-core/ctaes.git (branch master)
+* for `src/crc32c`: https://github.com/google/crc32c.git (branch master)
```
Comment (PastaPastaPasta) [inline test/lint/README.md:22]: test: Add crc32c to subtree check linter 
 
 doesn't do https://github.com/bitcoin/bitcoin/pull/17398/commits/84ff1b2076ef91ce688930d0aa0a7f4078ef3e1d#diff-a12151b9941bfec9b94144ff38616fb4e3c081bc96a25591f966eef045eafeecR17
 
 Why?

**Review by PastaPastaPasta — APPROVED** [review]
utACK
